server/src/main/java/org/opensearch/identity/PluginSubject.java

```diff
@@ -0,0 +1,36 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.identity;
+
+import org.opensearch.common.annotation.ExperimentalApi;
+
+import java.security.Principal;
+import java.util.concurrent.Callable;
+
+/**
+ * Similar to {@link Subject}, but represents a plugin executing actions
+ *
+ * @opensearch.experimental
+ */
+@ExperimentalApi
+public interface PluginSubject {
```

Comment:
`PluginSubject` is `Subject` , the only place I see `Subject::authenticate` is called  - RestController::handleAuthenticateUser so it seems like we need to have `UserSubject` extracted, does it make sense?

Reply:
yes that makes sense. I will create a hierarchy with Subject and 2 sub-interfaces: UserSubject and PluginSubject.

This actually brings me back to when I introduced persistent ThreadContext headers last year. Even if a PluginSubject calls runAs and executes transport actions as the plugin subject when handling an API request, the UserSubject should still be present and should be able to be retrieved. 

Pushing a change in a min.